### PR TITLE
core: useful error msgs

### LINKED ISSUE #106: Return meaningful errors from Core
https://godoc.org/google.golang.org/grpc/status

### PR DESCRIPTION
**Pull request by zhulia96**
:tickets: **Ticket(s)**: Closes #106

---

## :construction_worker: Changes
Added meaningful grpc errors for core

## :flashlight: Testing Instructions
Tested by adding new test cases in service_test.go

### CODE REVIEW

File: client/lib/api/index.test.js [hunk 1]
```diff
@@ -15,239 +15,253 @@ describe('API', () => {
...
-  describe('createAccountFail', () => {
-    test('ok', (done) => {
-      const a = new api.API();
-      moxios.stubRequest('/user/create', {
-        status: 300,
-        response: {},
+    const onFulfilled = sinon.spy();
+    a.getStatus().then(onFulfilled)
+      .then(() => {
+        expect(true).toBe(false);
```
Comment (leesw98) [inline client/lib/api/index.test.js]: Is this the correct way to be testing this?

File: core/service/service.go [hunk 2]
```diff
@@ -188,12 +188,18 @@ func (s *Service) CreateAccount(ctx context.Context, req *request.CreateAccount)
 		&models.User{Email: req.Email, Name: req.Name, Hash: salt},
 		&models.EmailVerification{Email: req.Email, Hash: v.Hash, Expiry: v.Expiry},
 	); err != nil {
-		return nil, fmt.Errorf("failed to create user: %s", err.Error())
+		if strings.Contains(err.Error(), "Keys cannot be empty") {
+			return nil, status.Errorf(codes.InvalidArgument, "failed to create user: %s", err.Error())
```
Comment (leesw98) [inline core/service/service.go]: 👍

File: client/lib/api/index.test.js [hunk 3]
```diff
@@ -15,239 +15,253 @@ describe('API', () => {
...
-      const a = new api.API();
-      moxios.stubRequest('/status', {
-        status: 200,
-        response: {
-          resp: 'active',
-        },
-      });
+describe('getStatus', () => {
+  test('ok', (done) => {
+    const a = new api.API();
```
Comment (leesw98) [inline client/lib/api/index.test.js]: If `new api.API()` and `sinon.spy()` are always being created, consider putting them in the `beforeEach`

File: client/lib/api/index.test.js (hunk 1 above)
Comment (zhulia96) [reply client/lib/api/index.test.js]: @bobheadxi what do you think?

Comment (bobheadxi) [reply client/lib/api/index.test.js]: @leesw98 the intention here is to explicitly fail the test. @zhulia96 maybe take a look to see if the testing library (Jest https://jestjs.io/) offers a way to fail a test with a useful error message?

Comment (bobheadxi) [reply client/lib/api/index.test.js]: i think i put this here originally because i couldnt find one :/

File: core/service/service.go [hunk 4]
```diff
@@ -215,10 +227,13 @@ func (s *Service) Verify(ctx context.Context, req *request.Verify) (*response.Me
...
+		if strings.Contains(err.Error(), "Email can not be empty") {
+			return nil, status.Errorf(codes.InvalidArgument, "failed to authenticate user: %s", err.Error())
+		}
+		return nil, status.Errorf(codes.Internal, "failed to authenticate user: %s", err.Error())
 	}
 	if crypto.ComparePasswords(user.Hash, req.GetPassword()) {
 		return &response.Message{Message: "user successfully logged in"}, nil
 	}
-	return nil, errors.New("user not authenticated")
+	return nil, status.Errorf(codes.InvalidArgument, "failed to authenticate user: %s", "passwords did not match")
```
Comment (bobheadxi) [inline core/service/service.go]: three things here:

* dont use `Errorf` and `%s` if the injected string isnt a variable, just use `Error` and make the string one string 😅 
* if you fail to authenticate, [Unauthenticated](https://github.com/grpc/grpc-go/blob/master/codes/codes.go#L146:2) might be a more suitable code
* telling the user specifically that `passwords did not match` is actually bad practice - it means that you've told them for sure that an account with the given username exists, and thus taken away half the knowledge they need to break into an account. That said, this is `core` and not `gateway`, so probably fine to say so here - we just need to make sure `gateway` *does not*

File: core/service/service.go [hunk 5]
```diff
@@ -188,12 +188,18 @@ func (s *Service) CreateAccount(ctx context.Context, req *request.CreateAccount)
 		&models.User{Email: req.Email, Name: req.Name, Hash: salt},
 		&models.EmailVerification{Email: req.Email, Hash: v.Hash, Expiry: v.Expiry},
 	); err != nil {
-		return nil, fmt.Errorf("failed to create user: %s", err.Error())
+		if strings.Contains(err.Error(), "Keys cannot be empty") {
```
Comment (bobheadxi) [inline core/service/service.go]: is this one of our errors? Go errors should be lowercase for consistency

```suggestion
		if strings.Contains(err.Error(), "keys cannot be empty") {
```

might want to update the source of the error for this

File: core/service/service.go [hunk 6]
```diff
@@ -206,7 +212,13 @@ func (s *Service) CreateAccount(ctx context.Context, req *request.CreateAccount)
 func (s *Service) Verify(ctx context.Context, req *request.Verify) (*response.Message, error) {
 	v, err := s.db.GetEmailVerification(req.GetEmail(), req.GetHash())
 	if err != nil {
-		return nil, err
+		if strings.Contains(err.Error(), "Email or hash can not be empty") {
```
Comment (bobheadxi) [inline core/service/service.go]: same here

```suggestion
		if strings.Contains(err.Error(), "email or hash can not be empty") {
```

might want to check the source

**Review by bobheadxi — COMMENTED** [review]
good stuff! pretty close to done for now i think - i left a few comments, and @leesw98 made a few good points too 👍 good work! 🚀

**Review by bobheadxi — APPROVED** [review]
gucci!

**Pull request merged**